### PR TITLE
Add a GDScript cache singleton

### LINKED ISSUE #21461: Some usages of `class_name` can produce cyclic errors unexpectedly
___

***Bugsquad edit:** This issue has been confirmed several times already. No need to confirm it further.*

___

Windows 10 64-bit - Godot 3.1 alpha

## Edit - Refined issue to focus on cyclic errors.

Certain uses of `class_name` can produce cyclic errors in situations where there are no cyclic references, and in fact sometimes no other references outside of the single one given.

I've noticed this most frequently in tool script when using `is` to make a class comparison, but it seems to happen at times in normal script situations.

Recreating it in a sample project has eluded me.

godot master_2018-08-27_00-47-02

### LINKED ISSUE #27136: Classes cross (recursive) references lead to "script could not be loaded"
**Godot version:** 3.1

**OS/device including version:** Ubuntu 18.04

**Issue description:**
Say, you have the following:
```gdscript
# one.gd
extends Node2D
class_name One

func use_two(var two: Two):
	pass
```

```gdscript
# two.gd
extends Node
class_name Two

func use_one(var one: One):
	pass
```
The project will fail on startup with error "Class 'One' was found in global scope but its script could not be loaded"

If we we remove one of type annotations `: One` or  `: Two` from functions above, it will launch without any complaint

**Minimal reproduction project:**
[repro_recursive.tar.gz](https://github.com/godotengine/godot/files/2974313/repro_recursive.tar.gz)

### PR DESCRIPTION
**Pull request by vnen**
This singleton is able to cache partially compiled scripts and is independent of the resource loader, which allows cyclic dependencies to be properly resolved.

There's still one problem here: stuff never get out of the cache. I thought there was a safe place to clear it but I was wrong, so it needs a better solution. If someone wants to contribute a way to do it safely I will be glad, but I've been working in this too long that isn't good for my head to keep thinking on it, so I'll pass it on to the community.

I've been testing this with a real-world project, but I think it needs some more testing to be confident it's working properly.

Fix #21461
Fix #27136

### CODE REVIEW

**Comment by Zylann** [comment]
Does this also apply to `preload`? I exclusively use that to import classes inside plugins, instead of global class names

**Comment by vnen** [comment]
> Does this also apply to `preload`? I exclusively use that to import classes inside plugins, instead of global class names

No, because `preload` calls the resource loader directly, since it can load other resource types, so it's not possible to have loading cycles. I guess we could detect when you are preloading a script (since path must be constant) and use the cache instead, but I'm not sure.

File: modules/gdscript/gdscript_parser.h [hunk 1]
```diff
@@ -42,6 +42,26 @@ struct GDScriptWarning;
...
+		bool is_mixed(IndentLevel other) {
+			return (
+					(indent == other.indent && tabs != other.tabs) ||
+					(indent > other.indent && tabs < other.tabs) ||
+					(indent < other.indent && tabs > other.tabs));
+		}
+
+		IndentLevel() :
+				indent(0),
+				tabs(0) {}
```
Comment (akien-mga) [inline modules/gdscript/gdscript_parser.h:58]: You can initialize directly upon declaration.

Comment (akien-mga) [reply modules/gdscript/gdscript_parser.h:58]: Ah well this is the `3.2` branch so it's not so important, we hadn't ported the codebase yet.

**Comment by vnen** [comment]
Leave as draft since it likely can't be merged as it is. I'll come back to this once it's working on master.

**Comment by Xrayez** [comment]
Is this one of the PRs which could help #27333 and #34205 in the future? I currently come up with a solution to solve this in 3.2 in #39797, and I think it does resemble similar functionality (just asking whether it makes sense to make #39797 work against 4.0).

**Comment by vnen** [comment]
@Xrayez I don't have these issues in mind with this PR. Though the cache singleton could potentially be used to cache those as well.

**Comment by Xrayez** [comment]
Sorry for polluting with test results not related to the issues this PR aims to fix, but I'm reporting that this PR doesn't solve the performance issues with #27333 nor #34205. I think this might provide some performance boost but #39793 fixes some major bottlenecks already which was cherry-picked to 3.2.2, so the performance boost is "shadowed".

> Though the cache singleton could potentially be used to cache those as well.

Yeah, that was my initial assumption of what kind of issues this PR could solve as a byproduct.

Throughout my research on this topic, I've figured that #39797 provides the ultimate solution, but yet to be tested as well.

**Comment by akien-mga** [comment]
> This a dead PR? Came here from the blog post.

What blog post?

**Comment by atirut-w** [comment]
https://godotengine.org/article/gdscript-progress-report-type-checking-back#:~:text=GDScript%20cache%20singleton%20for%20mediating%20loading%20of%20scripts

**Comment by lbfalvy** [comment]
> This a dead PR? Came here from the blog post.

I think this PR became obsolete with the GDScript rewrite and two-stage resource loader a while back. That project had regular reports on the Godot blog for a year but I'm not sure whether it got into any current release because the series shifted my interest from gamedev to langdev.

**Comment by atirut-w** [comment]
> > This a dead PR? Came here from the blog post.
> 
> I think this PR became obsolete with the GDScript rewrite and two-stage resource loader a while back. That project had regular reports on the Godot blog for a year but I'm not sure whether it got into any current release because the series shifted my interest from gamedev to langdev.

Wasn't the GDScript rewrite for 4.0?

**Comment by lbfalvy** [comment]
> > > This a dead PR? Came here from the blog post.
> > 
> > 
> > I think this PR became obsolete with the GDScript rewrite and two-stage resource loader a while back. That project had regular reports on the Godot blog for a year but I'm not sure whether it got into any current release because the series shifted my interest from gamedev to langdev.
> 
> Wasn't the GDScript rewrite for 4.0?

Yeah it was, my bad. The blog post series that you linked is talking exactly about that rewrite though so I guess this is a dead PR for non-dead functionality that will be merged into 4.0.

**Comment by atirut-w** [comment]
> I guess this is a dead PR for non-dead functionality that will be merged into 4.0.

That's sad. This would be a great thing to have in 3.x

**Comment by vnen** [comment]
This could still work but I don't have the bandwidth to finish this while also fixing things for 4.0. Even the cache done in `master` still has some issues.

For now I'll keep this closed since it's unlikely I'll have the opportunity to get back to it.

**Pull request closed without merging**